gzip/gzip.go

```diff
@@ -0,0 +1,83 @@
+package gzip
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io/ioutil"
+
+	kafka "github.com/segmentio/kafka-go"
+)
+
+func init() {
+	kafka.RegisterCompressionCodec(func() kafka.CompressionCodec {
+		return NewCompressionCodec()
+	})
+}
+
+type CompressionCodec struct {
+	CompressionLevel int
+}
+
+func NewCompressionCodec() CompressionCodec {
+	return NewCompressionCodecWith(kafka.DefaultCompressionLevel)
+}
+
+func NewCompressionCodecWith(level int) CompressionCodec {
+	return CompressionCodec{
+		CompressionLevel: level,
+	}
+}
+
+// Code implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) Code() int8 {
+	return 1
+}
+
+// Encode implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) Encode(dst, src []byte) (int, error) {
+	var buf bytes.Buffer
+	writer := gzip.NewWriter(&buf)
```

Comment:
Creating a gzip writer is not a cheap operation, the constructor has to allocate multiple buffers internally, which is why I was suggesting you to keep the writer as a field of the compression codec, and use its `Reset` method to set the writer that it will be outputting to compressed data to.